sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java

```diff
@@ -502,6 +511,103 @@ public void testKafkaWithDelayedStopReadingFunction() {
+                                    .addValue(new Random(numb).nextDouble())
+                                    .addValue(
+                                        Lists.newArrayList(
+                                            Long.toString(numb + 1),
+                                            Long.toString(numb + 2))) // childrenNames
+                                    .build()))
+                .setRowSchema(KAFKA_TOPIC_SCHEMA))
+        .apply(
+            "Write to Kafka",
+            new KafkaWriteSchemaTransformProvider()
+                .from(
+                    KafkaWriteSchemaTransformProvider.KafkaWriteSchemaTransformConfiguration
+                        .builder()
+                        .setTopic(options.getKafkaTopic() + "-schema-transform")
+                        .setBootstrapServers(options.getKafkaBootstrapServerAddresses())
+                        .setFormat(format)
+                        .build())
+                .buildTransform());
+
+    PAssert.that(
+            PCollectionRowTuple.empty(readPipeline)
+                .apply(
+                    "Read from unbounded Kafka",
+                    // A timeout of 30s for local, container-based tests, and 2 minutes for
+                    // real-kafka tests.
+                    new KafkaReadSchemaTransformProvider(
+                            true, options.isWithTestcontainers() ? 30 : 120)
+                        .from(
+                            KafkaReadSchemaTransformConfiguration.builder()
+                                .setDataFormat(format)
+                                .setAutoOffsetResetConfig("earliest")
+                                .setSchema(schemaDefinition)
+                                .setTopic(options.getKafkaTopic() + "-schema-transform")
+                                .setBootstrapServers(options.getKafkaBootstrapServerAddresses())
+                                .build())
+                        .buildTransform())
+                .get("output"))
+        .containsInAnyOrder(
+            LongStream.range(0L, 1000L)
+                .<Row>mapToObj(
+                    numb ->
+                        Row.withSchema(KAFKA_TOPIC_SCHEMA)
+                            .addValue(Long.toString(numb)) // Name
+                            .addValue(Long.valueOf(Long.valueOf(numb).hashCode())) // User ID
+                            .addValue(Long.valueOf(numb).intValue()) // Age
+                            .addValue(numb % 2 == 0) // ageIsEven
+                            .addValue(new Random(numb).nextDouble())
+                            .addValue(
+                                Lists.newArrayList(
+                                    Long.toString(numb + 1),
+                                    Long.toString(numb + 2))) // childrenNames
+                            .build())
+                .collect(Collectors.toList()));
+
+    PipelineResult writeResult = writePipeline.run();
+    writeResult.waitUntilFinish();
```

Comment:
Shall we assert pipeline status (not fail)? Same for the read pipeline below.

Reply:
good idea. done!